docs/ops/state/checkpoints.md

```diff
@@ -99,3 +99,296 @@ above).
+
+Notice that, unlike in full checkpoints, we also must record changes that delete state in an incremental checkpoint, as
+in the case of ``K0``. In this simple example, we can see how incremental checkpointing can reduce the amount of data that
+is written for each checkpoint.
+
+The next interesting question: how does restoring from incremental checkpoints compare to restoring from full
+checkpoints? Restoring a full checkpoint is as simple as loading all the data from the checkpoint back into the job
+state because full checkpoints are self-contained. In contrast, to restore an incremental checkpoint, we need to replay
+the history of all incremental checkpoints that are in the reference chain of the checkpoint we are trying to restore.
+In our example from Figure 1, those connections are represented by the orange arrows. If we want to restore ``CP 3``, as a
+first step, we need to apply all the changes of ``CP 1`` to the empty initial job state. On top of that, we apply the
+changes from ``CP 2``, and then the changes from ``CP 3``.
+
+A different way to think about basic incremental checkpoints is to imagine it as a changelog with some aggregation. What
+we mean by aggregated is that for example, if the state under key ``K1`` is overwritten multiple times between two
+consecutive checkpoints, we will only record the latest state value at the time in the checkpoint. All previous changes
+are thereby subsumed.
+
+This leads us to the discussion of the potential *disadvantages* of incremental checkpoints compared to full checkpoints.
+While we save work in writing checkpoints, we have to do more work in reading the data from multiple checkpoints on
+recovery. Furthermore, we can no longer simply delete old checkpoints because new checkpoints rely upon them and the
+history of old checkpoints can grow indefinitely over time (like a changelog).
+
+At this point, it looks like we didn’t gain too much from incremental checkpoints because we are again trading between
+checkpointing overhead and recovery time. Fortunately, there are ways to improve on this naive approach to recovery. One
+simple and obvious way to restrict recovery time and the length of the checkpoint history is to write a full checkpoint
+from time to time. We can drop all checkpoints prior to the most recent full checkpoint, and the full checkpoint can
+serve as a new basis for future incremental checkpoints.
+
+Our actual implementation of incremental checkpoints in Flink is more involved and designed to address a number of
+different tradeoffs. Our incremental checkpointing restricts the size of the checkpoint history and therefore never
+needs take a full checkpoint to keep recovery efficiently! We present more detail about this in the next section, but
+the high level idea is to accept a small amount of redundant state writing to incrementally introduce
+merged/consolidated replacements for previous checkpoints. For now, you can think about Flink’s approach as stretching
```

Comment:
the high level idea is to accept a small amount of redundant state writing that incrementally introduces
merged/consolidated replacements for previous checkpoints.